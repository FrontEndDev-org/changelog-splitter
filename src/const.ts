export const pkgName = process.env.PKG_NAME!;
export const pkgVersion = process.env.PKG_VERSION!;

export const generatedSeparator = `<!-- Generated by ${pkgName}@v${pkgVersion}, please do not modify manually!! -->`;
export const generatedSeparatorRE = new RegExp(`<!-- .*?${pkgName}@v.*? -->`);
