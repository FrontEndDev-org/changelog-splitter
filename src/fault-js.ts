export type FaultDefinition = Record<string, string>;

// export interface Error

// export type ExtractFault =

export function defineFault<F extends FaultDefinition, C extends keyof F>(name: string, definition: F) {
  return class Fault extends Error {
    readonly name = name;
    readonly code: C;

    constructor(code: C, message?: string) {
      super(definition[code]);
      this.code = code;
      if (message) this.message = message;

      // @ref https://stackoverflow.com/a/32749533
      if (typeof Error.captureStackTrace === 'function') {
        Error.captureStackTrace(this, this.constructor);
      } else {
        this.stack = new Error(message).stack;
      }
    }

    because(cause: Error) {
      this.cause = cause;
      return this;
    }
  };
}
