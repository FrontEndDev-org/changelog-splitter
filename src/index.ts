export { run } from './commands';
export { defineConfig } from './config';
export { splitChangelog } from './splitter';
