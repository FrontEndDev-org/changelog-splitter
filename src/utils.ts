import fs from 'fs';
import os from 'os';
import path from 'path';
import readline from 'readline';
import * as crypto from 'crypto';
import { pkgName, pkgVersion } from './const';
import MultiStream from 'multistream';

/**
 * 匹配版本号中的主版本号
 * @param {string} version
 * @returns {string}
 */
export function matchMajor(version: string) {
  return version.split('.')[0].replace(/\D/g, '');
}

/**
 * 从 markdown 中匹配版本标题，例如
 * "## [12.34.56](http://example.com) A new version (2022-11-22)"
 * @param {string} line
 * @returns {{date: string, major: string, name: string, link: string, title: string}}
 */
export function matchVersion(line: string) {
  // ## [version](link) title (date)
  // ## [version](link) (date)
  // ## version title (date)
  // ## version (date)
  const matches = line.match(/^#+\s+(\S+)(.*?)(?:\((.*?)\)|$)/);

  if (!matches) return;

  const nameLink = matches[1].match(/\[(.*?)]\((.*?)\)/);
  const name = nameLink ? nameLink[1] : matches[1];
  const major = matchMajor(name);

  if (!major) return;

  return {
    name,
    major,
    link: nameLink ? nameLink[2] : '',
    title: matches[2].trim(),
    date: matches[3] || '',
  };
}

/**
 * 从 markdown 中匹配前版本链接，如：
 * - [v1.x](changelogs/v1.x.md)
 * @param {string} line
 * @returns {{major: string, name: string, link: string}}
 */
export function matchPrevious(line: string) {
  // - [version](file)
  const matches = line.match(/\[(.*?)]\((.*?)\)/);

  if (!matches) return;

  const name = matches[1];
  const major = matchMajor(name);

  return {
    name,
    major,
    link: matches[2],
  };
}

/**
 * 按行读取文件内容
 * @param {string} filePath
 * @param {(line: string) => any} lineProcessor
 * @returns {Promise<void>}
 */
export async function readFileLineByLine(filePath: string, lineProcessor: (line: string) => any) {
  const inputStream = fs.createReadStream(filePath);
  const rl = readline.createInterface({
    input: inputStream,
    crlfDelay: Infinity,
  });

  for await (const line of rl) {
    await lineProcessor(line);
  }
}

/**
 * 统计文件的行数
 * @param {string} filePath
 * @returns {Promise<number>}
 */
export async function countFileLines(filePath: string): Promise<number> {
  let count = 0;
  await readFileLineByLine(filePath, () => count++);
  return count;
}

/**
 * 流式复制文件，会覆盖目标文件
 * @param {string} source
 * @param {string} target
 * @returns {Promise<unknown>}
 */
export async function pipeFile(source: string, target: string) {
  return new Promise<void>((resolve, reject) => {
    if (source === target) return resolve();

    const rs = fs.createReadStream(source);
    const ws = fs.createWriteStream(target);
    rs.pipe(ws).on('error', reject).on('close', resolve);
  });
}

/**
 * 合并多个文件
 * @param {string[]} files
 * @param {string} target
 * @returns {Promise<void>}
 */
export async function mergeFiles(files: string[], target: string) {
  return new Promise<void>((resolve, reject) => {
    const inputs = files.map((file) => fs.createReadStream(file));
    const output = fs.createWriteStream(target);
    new MultiStream(inputs).pipe(output).on('error', reject).on('close', resolve);
  });
}

/**
 * 创建临时目录【必存在】
 * @returns {string}
 */
export function createTempDirname() {
  const d = path.join(os.tmpdir(), pkgName, pkgVersion, crypto.randomUUID() + '.d');
  fs.mkdirSync(d, { recursive: true });
  return d;
}

/**
 * 创建临时文件【必存在】
 * @param {string} data
 * @returns {string}
 */
export function createTempFile(data = '') {
  const d = createTempDirname();
  const f = path.join(d, crypto.randomUUID() + '.f');
  fs.writeFileSync(f, data);
  return f;
}

/**
 * 根据主版本号生成实际字符串
 * @param {string} templateName
 * @param {string} major
 * @returns {string}
 */
export function generateNameByMajor(templateName: string, major: string) {
  return templateName.replace(/\[major]/g, major);
}
