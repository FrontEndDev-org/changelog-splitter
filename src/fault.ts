import { defineFault } from './fault-js';

const ConfigFaultDefinition = {
  CHANGELOG_NOT_FOUND: 'CHANGELOG 文件不存在',
  PACKAGE_JSON_READ_ERROR: 'package.json 文件读取失败',
  PACKAGE_JSON_PARSE_ERROR: 'package.json 文件解析失败',
};
export const ConfigFault = defineFault('ConfigFault', ConfigFaultDefinition);
export type ConfigFaultInstance = InstanceType<typeof ConfigFault>;
export type ConfigFaultCode = keyof typeof ConfigFaultDefinition;
