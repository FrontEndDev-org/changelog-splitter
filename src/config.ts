import fs from 'fs';
import path from 'path';
import { ExtendConfig, RuntimeConfig, StrictUserConfig, UserConfig } from './types';
import { createTempDirname, matchMajor } from './utils';

export const defaults: StrictUserConfig = {
  cwd: process.cwd(),
  currentChangelogFile: 'CHANGELOG.md',
  packageFile: 'package.json',
  previousVersionChangelogFileName: 'changelogs/v[major].x.md',
  previousVersionChangelogTitle: '# v[major].x 更新日志',
  previousVersionLinkTitle: '## 其他版本的更新日志',
  processVersion: () => 0,
};

/**
 * 定义配置
 * @param {UserConfig} config
 * @returns {StrictUserConfig}
 */
export function defineConfig(config?: UserConfig): StrictUserConfig {
  return Object.assign({}, defaults, config) as StrictUserConfig;
}

/**
 * 创建运行期配置
 * @param {StrictUserConfig} strictUserConfig
 * @returns {RuntimeConfig}
 */
export function createRuntimeConfig(strictUserConfig: StrictUserConfig): RuntimeConfig {
  const { cwd, packageFile, currentChangelogFile } = strictUserConfig;
  const resolvePath = (...segments: string[]): string => path.resolve(cwd, ...segments);
  const currentChangelogFilePath = resolvePath(currentChangelogFile);
  const currentMajorChangelogFilePath = path.join(createTempDirname(), `CHANGELOG.md`);

  const packageFilePath = resolvePath(packageFile);
  const json = fs.readFileSync(packageFilePath, 'utf8');
  const { version: currentVersion } = JSON.parse(json) as { version: string };
  const currentMajor = matchMajor(currentVersion);

  return Object.assign<any, any, ExtendConfig>({}, strictUserConfig, {
    resolvePath,
    currentChangelogFilePath,
    currentVersion,
    currentMajor,
    currentMajorChangelogFilePath,
  });
}
