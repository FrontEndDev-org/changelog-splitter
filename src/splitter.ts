import * as fs from 'fs';
import * as path from 'path';
import { ConflictStrategy, createRuntimeConfig, RuntimeConfig, StrictUserConfig } from './config';
import { generatedSeparator, generatedSeparatorRE } from './const';
import { SplitFault } from './fault';
import {
  createTempFile,
  generateNameByMajor,
  matchPrevious,
  matchVersion,
  mergeFiles,
  pipeFile,
  readFileLineByLine,
} from './utils';

export interface SplitResult {
  runtimeConfig: RuntimeConfig;

  /**
   * 已经处理的主版本号与文件的映射表
   */
  processedFileByMajor: { [major: string]: string };

  /**
   * 主版本更新日志末尾的连续空白长度
   */
  blankLengthByMajor: { [major: string]: number };

  /**
   * 由于冲突策略导致的废弃文件
   */
  deprecatedMajorFiles: { [major: string]: string };
}

/**
 * 分离当前更新日志
 * @param {RuntimeConfig} runtimeConfig
 * @returns {Promise<void>}
 */
export async function splitCurrentChangelog(runtimeConfig: RuntimeConfig): Promise<SplitResult> {
  const {
    previousVersionChangelogTitle,
    previousVersionChangelogFileName,
    previousVersionChangelogConflictStrategy,
    resolvePath,
    currentChangelogFilePath,
    currentMajor,
    currentVersionChangeTempFilePath,
  } = runtimeConfig;

  const splitResult: SplitResult = {
    runtimeConfig,
    processedFileByMajor: {},
    blankLengthByMajor: {},
    deprecatedMajorFiles: {},
  };
  const { processedFileByMajor, blankLengthByMajor, deprecatedMajorFiles } = splitResult;

  const process = (major: string, line: string) => {
    const isCurrentMajor = major === currentMajor;
    const filePath = isCurrentMajor
      ? currentVersionChangeTempFilePath
      : resolvePath(generateNameByMajor(previousVersionChangelogFileName, major));

    fs.mkdirSync(path.dirname(filePath), { recursive: true });

    // 第一次处理，清空其本身内容
    if (!processedFileByMajor[major]) {
      fs.writeFileSync(filePath, '');
    }

    // 未处理过的大版本 && 不是当前版本 = 第一次处理旧版本更新日志
    if (!processedFileByMajor[major] && !isCurrentMajor) {
      fs.writeFileSync(filePath, `${generatedSeparator}\n\n`);

      // 旧版本标题
      if (previousVersionChangelogTitle) {
        const titleText = generateNameByMajor(previousVersionChangelogTitle, major);
        fs.appendFileSync(filePath, titleText + '\n\n');
      }
    }

    const blankLength = blankLengthByMajor[major] || 0;
    const isBlank = line.trim() === '';

    if (blankLength < 3 || !isBlank) {
      fs.appendFileSync(filePath, line + '\n');
    }

    blankLengthByMajor[major] = isBlank ? blankLength + 1 : 0;
    processedFileByMajor[major] = filePath;
  };

  let processingMajor = '';
  let processingPrevious = false;

  // title
  // version
  // ...
  // version
  // previousLink
  // ...
  // previousLink

  await readFileLineByLine(currentChangelogFilePath, async (line) => {
    const isPreviousBlock = generatedSeparatorRE.test(line);

    // 前版本块
    if (isPreviousBlock) {
      processingPrevious = true;
      return;
    }

    // 处理前版本引用链接，此时已经处理过更新日志中的所有版本了
    if (processingPrevious) {
      const previous = matchPrevious(line);

      if (!previous) return;

      const { link, major } = previous;
      // 旧文件
      const processedFile = resolvePath(link);
      // 新文件
      const processingFile =
        processedFileByMajor[major] || resolvePath(generateNameByMajor(previousVersionChangelogFileName, major));
      // 是否存在两份旧版本文件
      const conflicting = processedFile !== processingFile;

      if (!conflicting) return;

      const existProcessingFile = fs.existsSync(processingFile);
      const existProcessedFile = fs.existsSync(processedFile);

      // 新旧文件同时存在：合并文件
      const acceptedFile =
        previousVersionChangelogConflictStrategy === ConflictStrategy.ProcessedFile
          ? (processedFileByMajor[major] = processedFile)
          : (processedFileByMajor[major] = processingFile);

      if (existProcessingFile && existProcessedFile) {
        const tempFile = createTempFile();
        await mergeFiles(
          [
            // 旧文件在前
            processedFile,
            // 新文件在后
            processingFile,
          ],
          tempFile
        );
        await pipeFile(tempFile, acceptedFile);
        if (processedFile !== acceptedFile) deprecatedMajorFiles[major] = processedFile;
      }
      // 只有旧文件
      else if (existProcessedFile) {
        await pipeFile(processedFile, acceptedFile);
        if (processedFile !== acceptedFile) deprecatedMajorFiles[major] = processedFile;
      } else {
        throw new SplitFault('LINK_CHANGELOG_NOT_FOUND', `链接的更新日志文件不存在 ${processedFile}`);
      }

      return;
    }

    const version = matchVersion(line);

    // 版本开始
    if (version) {
      processingMajor = version.major;
      process(processingMajor, line);
    }
    // 版本区
    else if (processingMajor) {
      process(processingMajor, line);
    }
    // 标题块
    else {
      process(currentMajor, line);
    }
  });

  return splitResult;
}

/**
 * 引用之前的更新日志链接
 * @param {RuntimeConfig} runtimeConfig
 * @param {SplitResult} splitResult
 */
export async function referPreviousChangelog(runtimeConfig: RuntimeConfig, splitResult: SplitResult) {
  const { currentVersionChangeTempFilePath, currentMajor, previousVersionLinkTitle, currentVersionChangeFilePath } =
    runtimeConfig;

  const { processedFileByMajor, blankLengthByMajor } = splitResult;
  const prevVersions = Object.keys(processedFileByMajor)
    .filter((v) => v !== currentMajor)
    .map((v) => parseInt(v, 10))
    .sort((a, b) => b - a);

  // 需要链接其他版本
  if (prevVersions.length > 0) {
    // 通常是存在的，为了便于单元测试时不存在
    if (!fs.existsSync(currentVersionChangeTempFilePath)) {
      fs.mkdirSync(path.dirname(currentVersionChangeTempFilePath), { recursive: true });
    }

    if (!blankLengthByMajor[currentMajor]) {
      fs.appendFileSync(currentVersionChangeTempFilePath, `\n\n`);
    }

    fs.appendFileSync(currentVersionChangeTempFilePath, `${generatedSeparator}\n\n`);
    fs.appendFileSync(currentVersionChangeTempFilePath, `${previousVersionLinkTitle}\n`);

    const currentDir = path.dirname(currentVersionChangeFilePath);
    prevVersions.forEach((v) => {
      const filePath = processedFileByMajor[v];
      const relativePath = path.relative(currentDir, filePath);
      fs.appendFileSync(currentVersionChangeTempFilePath, `- [v${v}.x](${relativePath})\n`);
    });

    fs.appendFileSync(currentVersionChangeTempFilePath, '\n');
  }

  // 通常是存在的，为了便于单元测试时不存在
  if (fs.existsSync(currentVersionChangeTempFilePath)) {
    // 复制当前大版本的更新日志临时文件回原地
    await pipeFile(currentVersionChangeTempFilePath, currentVersionChangeFilePath);
    fs.rmSync(currentVersionChangeTempFilePath);
  }
}

export async function splitChangelog(config: StrictUserConfig) {
  const runtimeConfig = createRuntimeConfig(config);
  const result = await splitCurrentChangelog(runtimeConfig);
  await referPreviousChangelog(runtimeConfig, result);
  return result;
}
