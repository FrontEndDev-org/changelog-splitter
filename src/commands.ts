import chalk from 'chalk';
import process from 'process';
import { defineConfig } from './config';
import { parseCurrentChangelog } from './splitter';

export function run() {
  // const config = defineConfig();
  // splitCurrentChangelog(config)
  //   .then(() => {
  //     process.exit(0);
  //   })
  //   .catch((error) => {
  //     console.log(chalk.redBright(error.message));
  //     process.exit(1);
  //   });
}
