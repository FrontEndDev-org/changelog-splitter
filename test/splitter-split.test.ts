import fs from 'fs';
import path from 'path';
import { expect, test, describe } from 'vitest';
import { defineConfig } from '../src';
import { createRuntimeConfig } from '../src/config';
import { splitCurrentChangelog, SplitResult } from '../src/splitter';
import { createTempDirname } from '../src/utils';
import { ChangelogFolder, makeChangelogCwd } from './helpers';

// test/changelogs
const testChangelogsPath = path.join(__dirname, 'changelogs');

function readSplitResultFile(splitResult: SplitResult, major: string) {
  return fs.readFileSync(splitResult.processedFileByMajor[major], 'utf8');
}

test(ChangelogFolder.EmptyVersions, async () => {
  const cwd = makeChangelogCwd(ChangelogFolder.EmptyVersions);
  const config = createRuntimeConfig(
    defineConfig({
      cwd,
    })
  );
  const currentChangelogOrigin = fs.readFileSync(config.currentChangelogFilePath, 'utf8');
  const result = await splitCurrentChangelog(config);
  const currentChangelogNow = fs.readFileSync(config.currentChangelogFilePath, 'utf8');

  expect(currentChangelogOrigin).toEqual(currentChangelogNow);
  expect(Object.keys(result.processedFileByMajor)).toEqual([]);
});

test(ChangelogFolder.OnlyCurrentVersion, async () => {
  const cwd = makeChangelogCwd(ChangelogFolder.OnlyCurrentVersion);
  const config = createRuntimeConfig(
    defineConfig({
      cwd,
    })
  );
  const currentChangelogOrigin = fs.readFileSync(config.currentChangelogFilePath, 'utf8');
  const result = await splitCurrentChangelog(config);
  const currentChangelogNow = fs.readFileSync(config.currentChangelogFilePath, 'utf8');

  expect(currentChangelogOrigin).toEqual(currentChangelogNow);
  expect(Object.keys(result.processedFileByMajor)).toEqual(['2']);
  expect(readSplitResultFile(result, '2')).toMatchSnapshot();
});

test(ChangelogFolder.NotCurrentVersion, async () => {
  const cwd = makeChangelogCwd(ChangelogFolder.NotCurrentVersion);
  const config = createRuntimeConfig(
    defineConfig({
      cwd,
    })
  );
  const result = await splitCurrentChangelog(config);

  expect(Object.keys(result.processedFileByMajor).sort()).toEqual(['2', '3']);
  expect(readSplitResultFile(result, '2')).toMatchSnapshot();
  expect(readSplitResultFile(result, '3')).toMatchSnapshot();
});

test(ChangelogFolder.HasManyVersions, async () => {
  const cwd = makeChangelogCwd(ChangelogFolder.HasManyVersions);
  const config = createRuntimeConfig(
    defineConfig({
      cwd,
    })
  );
  const result = await splitCurrentChangelog(config);

  expect(Object.keys(result.processedFileByMajor).sort()).toEqual(['14', '16', '2']);
  expect(readSplitResultFile(result, '14')).toMatchSnapshot();
  expect(readSplitResultFile(result, '16')).toMatchSnapshot();
  expect(readSplitResultFile(result, '2')).toMatchSnapshot();
});

test(ChangelogFolder.HasRefVersions, async () => {
  const cwd = makeChangelogCwd(ChangelogFolder.HasRefVersions);
  const config = createRuntimeConfig(
    defineConfig({
      cwd,
    })
  );
  const result = await splitCurrentChangelog(config);

  expect(Object.keys(result.processedFileByMajor).sort()).toEqual(['10', '14', '16', '2', '9']);
  expect(readSplitResultFile(result, '10')).toMatchSnapshot();
  expect(readSplitResultFile(result, '14')).toMatchSnapshot();
  expect(readSplitResultFile(result, '16')).toMatchSnapshot();
  expect(readSplitResultFile(result, '2')).toMatchSnapshot();
  expect(readSplitResultFile(result, '9')).toMatchSnapshot();

  expect(result.processedFileByMajor[9]).toEqual(path.join(config.cwd, '123.md'));
  expect(result.processedFileByMajor[10]).toEqual(path.join(config.cwd, '456.md'));
});
