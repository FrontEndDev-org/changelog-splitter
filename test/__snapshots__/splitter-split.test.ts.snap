// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2-only-current-version 1`] = `
"# Changelog

## [2.7.14](https://github.com/vuejs/vue/compare/v2.7.13...v2.7.14) (2022-11-09)

### Bug Fixes

a

## [2.7.13](https://github.com/vuejs/vue/compare/v2.7.12...v2.7.13) (2022-10-14)

b

## [2.7.12](https://github.com/vuejs/vue/compare/v2.7.11...v2.7.12) (2022-10-12)
"
`;

exports[`3-not-current-version 1`] = `
"<!-- Generated by pkg-name-for-test@vpkg-version-for-test, please do not modify manually!! -->

# v2.x 更新日志

## [2.7.14](https://github.com/vuejs/vue/compare/v2.7.13...v2.7.14) (2022-11-09)

### Bug Fixes

a

## [2.7.13](https://github.com/vuejs/vue/compare/v2.7.12...v2.7.13) (2022-10-14)

b

## [2.7.12](https://github.com/vuejs/vue/compare/v2.7.11...v2.7.12) (2022-10-12)
"
`;

exports[`3-not-current-version 2`] = `
"# Changelog

"
`;

exports[`4-has-many-versions 1`] = `
"<!-- Generated by pkg-name-for-test@vpkg-version-for-test, please do not modify manually!! -->

# v14.x 更新日志

# 14.2.10 (2022-11-09)

c

<!-- CHANGELOG SPLIT MARKER -->

"
`;

exports[`4-has-many-versions 2`] = `
"# Changelog

# 16.0.0-next.5 (2023-03-29)

## Breaking Changes

<!-- CHANGELOG SPLIT MARKER -->

# 16.0.0-next.4 (2023-03-22)

d
"
`;

exports[`4-has-many-versions 3`] = `
"<!-- Generated by pkg-name-for-test@vpkg-version-for-test, please do not modify manually!! -->

# v2.x 更新日志

## [2.7.14](https://github.com/vuejs/vue/compare/v2.7.13...v2.7.14) (2022-11-09)

### Bug Fixes

a

## [2.7.13](https://github.com/vuejs/vue/compare/v2.7.12...v2.7.13) (2022-10-14)

b

## [2.7.12](https://github.com/vuejs/vue/compare/v2.7.11...v2.7.12) (2022-10-12)

"
`;

exports[`5-has-ref-versions 1`] = `
"9.x changelog
"
`;

exports[`5-has-ref-versions 2`] = `
"<!-- Generated by pkg-name-for-test@vpkg-version-for-test, please do not modify manually!! -->

# v14.x 更新日志

# 14.2.10 (2022-11-09)

c

<!-- CHANGELOG SPLIT MARKER -->

"
`;

exports[`5-has-ref-versions 3`] = `
"# Changelog

# 16.0.0-next.5 (2023-03-29)

## Breaking Changes

<!-- CHANGELOG SPLIT MARKER -->

# 16.0.0-next.4 (2023-03-22)

d


"
`;

exports[`5-has-ref-versions 4`] = `
"<!-- Generated by pkg-name-for-test@vpkg-version-for-test, please do not modify manually!! -->

# v2.x 更新日志

## [2.7.14](https://github.com/vuejs/vue/compare/v2.7.13...v2.7.14) (2022-11-09)

### Bug Fixes

a

## [2.7.13](https://github.com/vuejs/vue/compare/v2.7.12...v2.7.13) (2022-10-14)

b

## [2.7.12](https://github.com/vuejs/vue/compare/v2.7.11...v2.7.12) (2022-10-12)

"
`;

exports[`5-has-ref-versions 5`] = `
"10.x changelog
"
`;
